File: app/src/main/appshell/java/com/qubee/messenger/transport/RelayProtocol.kt
package com.qubee.messenger.transport

import org.json.JSONArray
import org.json.JSONObject

sealed interface ParsedRelayFrame {
    data class Challenge(val relaySessionId: String, val challenge: String) : ParsedRelayFrame
    data class Authenticated(val relaySessionId: String, val handle: String) : ParsedRelayFrame
    data class Envelope(val envelope: RelayEnvelope) : ParsedRelayFrame
    data class DeliveryAck(val messageId: String, val deliveredAt: Long) : ParsedRelayFrame
    data class PeerBundle(val peerHandle: String, val publicBundleBase64: String?) : ParsedRelayFrame
    data class ContactRequest(val request: RelayContactRequest) : ParsedRelayFrame
    data class HistorySync(val sync: RelayHistorySync) : ParsedRelayFrame
    data class Receipt(val receipt: RelayReceipt) : ParsedRelayFrame
    data class ReadCursor(val cursor: RelayReadCursor) : ParsedRelayFrame
    data class Signaling(val fromHandle: String, val toHandle: String, val signalType: String, val payload: String, val sentAt: Long) : ParsedRelayFrame
    data class Error(val message: String) : ParsedRelayFrame
    data object Ignored : ParsedRelayFrame
}

object RelayProtocol {
    fun helloJson(hello: RelayHello): String = JSONObject()
        .put("type", "hello")
        .put("handle", hello.handle)
        .put("deviceId", hello.deviceId)
        .put("displayName", hello.displayName)
        .put("publicBundleBase64", hello.publicBundleBase64)
        .put("identityFingerprint", hello.identityFingerprint)
        .toString()

    fun authenticateJson(proof: RelayAuthProof): String = JSONObject()
        .put("type", "authenticate")
        .put("handle", proof.handle)
        .put("relaySessionId", proof.relaySessionId)
        .put("challenge", proof.challenge)
        .put("publicBundleBase64", proof.publicBundleBase64)
        .put("identityFingerprint", proof.identityFingerprint)
        .put("signatureBase64", proof.signatureBase64)
        .toString()

    fun publishJson(envelope: RelayEnvelope): String = JSONObject()
        .put("type", "publish")
        .put("envelope", envelope.toJson())
        .toString()

    fun contactRequestJson(request: RelayContactRequest): String = JSONObject()
        .put("type", "contact_request")
        .put("request", request.toJson())
        .toString()

    fun receiptJson(receipt: RelayReceipt): String = JSONObject()
        .put("type", "receipt")
        .put("receipt", receipt.toJson())
        .toString()

    fun readCursorJson(cursor: RelayReadCursor): String = JSONObject()
        .put("type", "read_cursor")
        .put("cursor", cursor.toJson())
        .toString()

    fun requestPeerBundleJson(peerHandle: String): String = JSONObject()
        .put("type", "peer_bundle_request")
        .put("peerHandle", peerHandle)
        .toString()

    fun requestHistorySyncJson(since: Long): String = JSONObject()
        .put("type", "history_sync_request")
        .put("since", since)
        .toString()

    fun signalingJson(fromHandle: String, toHandle: String, signalType: String, payload: String, sentAt: Long): String = JSONObject()
        .put("type", "signaling")
        .put("fromHandle", fromHandle)
        .put("toHandle", toHandle)
        .put("signalType", signalType)
        .put("payload", payload)
        .put("sentAt", sentAt)
        .toString()

    fun parse(frameText: String): ParsedRelayFrame {
        val json = JSONObject(frameText)
        return when (json.optString("type")) {
            "challenge" -> ParsedRelayFrame.Challenge(
                relaySessionId = json.getString("relaySessionId"),
                challenge = json.getString("challenge"),
            )
            "authenticated" -> ParsedRelayFrame.Authenticated(
                relaySessionId = json.getString("relaySessionId"),
                handle = json.getString("handle"),
            )
            "delivery_ack" -> ParsedRelayFrame.DeliveryAck(
                messageId = json.getString("messageId"),
                deliveredAt = json.getLong("deliveredAt"),
            )
            "envelope" -> ParsedRelayFrame.Envelope(json.getJSONObject("envelope").toEnvelope())
            "peer_bundle_response" -> ParsedRelayFrame.PeerBundle(
                peerHandle = json.getString("peerHandle"),
                publicBundleBase64 = json.optString("publicBundleBase64").takeIf { it.isNotBlank() },
            )
            "contact_request" -> ParsedRelayFrame.ContactRequest(json.getJSONObject("request").toContactRequest())
            "history_sync_response" -> ParsedRelayFrame.HistorySync(json.toHistorySync())
            "receipt" -> ParsedRelayFrame.Receipt(json.getJSONObject("receipt").toReceipt())
            "read_cursor" -> ParsedRelayFrame.ReadCursor(json.getJSONObject("cursor").toReadCursor())
            "signaling" -> ParsedRelayFrame.Signaling(
                fromHandle = json.getString("fromHandle"),
                toHandle = json.getString("toHandle"),
                signalType = json.getString("signalType"),
                payload = json.getString("payload"),
                sentAt = json.getLong("sentAt"),
            )
            "error" -> ParsedRelayFrame.Error(json.optString("message", "Relay error"))
            else -> ParsedRelayFrame.Ignored
        }
    }

    private fun RelayEnvelope.toJson(): JSONObject = JSONObject()
        .put("messageId", messageId)
        .put("conversationId", conversationId)
        .put("senderHandle", senderHandle)
        .put("recipientHandle", recipientHandle)
        .put("sessionId", sessionId)
        .put("ciphertextBase64", ciphertextBase64)
        .put("algorithm", algorithm)
        .put("sentAt", sentAt)
        .put("senderDeviceId", senderDeviceId)

    private fun RelayContactRequest.toJson(): JSONObject = JSONObject()
        .put("requestId", requestId)
        .put("senderHandle", senderHandle)
        .put("recipientHandle", recipientHandle)
        .put("senderDisplayName", senderDisplayName)
        .put("publicBundleBase64", publicBundleBase64)
        .put("identityFingerprint", identityFingerprint)
        .put("sentAt", sentAt)

    private fun RelayReceipt.toJson(): JSONObject = JSONObject()
        .put("receiptId", receiptId)
        .put("messageId", messageId)
        .put("conversationId", conversationId)
        .put("senderHandle", senderHandle)
        .put("recipientHandle", recipientHandle)
        .put("recipientDeviceId", recipientDeviceId)
        .put("receiptType", receiptType)
        .put("recordedAt", recordedAt)

    private fun RelayReadCursor.toJson(): JSONObject = JSONObject()
        .put("cursorId", cursorId)
        .put("conversationId", conversationId)
        .put("handle", handle)
        .put("deviceId", deviceId)
        .put("readThroughTimestamp", readThroughTimestamp)
        .put("recordedAt", recordedAt)

    private fun JSONObject.toEnvelope(): RelayEnvelope = RelayEnvelope(
        messageId = getString("messageId"),
        conversationId = getString("conversationId"),
        senderHandle = getString("senderHandle"),
        recipientHandle = getString("recipientHandle"),
        sessionId = getString("sessionId"),
        ciphertextBase64 = getString("ciphertextBase64"),
        algorithm = getString("algorithm"),
        sentAt = getLong("sentAt"),
        senderDeviceId = optString("senderDeviceId", ""),
    )

    private fun JSONObject.toContactRequest(): RelayContactRequest = RelayContactRequest(
        requestId = getString("requestId"),
        senderHandle = getString("senderHandle"),
        recipientHandle = getString("recipientHandle"),
        senderDisplayName = getString("senderDisplayName"),
        publicBundleBase64 = getString("publicBundleBase64"),
        identityFingerprint = getString("identityFingerprint"),
        sentAt = getLong("sentAt"),
    )

    private fun JSONObject.toReceipt(): RelayReceipt = RelayReceipt(
        receiptId = getString("receiptId"),
        messageId = getString("messageId"),
        conversationId = getString("conversationId"),
        senderHandle = getString("senderHandle"),
        recipientHandle = getString("recipientHandle"),
        recipientDeviceId = getString("recipientDeviceId"),
        receiptType = getString("receiptType"),
        recordedAt = getLong("recordedAt"),
    )

    private fun JSONObject.toReadCursor(): RelayReadCursor = RelayReadCursor(
        cursorId = getString("cursorId"),
        conversationId = getString("conversationId"),
        handle = getString("handle"),
        deviceId = getString("deviceId"),
        readThroughTimestamp = getLong("readThroughTimestamp"),
        recordedAt = getLong("recordedAt"),
    )

    private fun JSONObject.toHistorySync(): RelayHistorySync = RelayHistorySync(
        relaySessionId = getString("relaySessionId"),
        syncedUntil = getLong("syncedUntil"),
        envelopes = getJSONArray("envelopes").toEnvelopeList(),
        contactRequests = getJSONArray("contactRequests").toContactRequestList(),
        receipts = optJSONArray("receipts")?.toReceiptList().orEmpty(),
        readCursors = optJSONArray("readCursors")?.toReadCursorList().orEmpty(),
    )

    private fun JSONArray.toEnvelopeList(): List<RelayEnvelope> = buildList {
        for (index in 0 until length()) add(getJSONObject(index).toEnvelope())
    }

    private fun JSONArray.toContactRequestList(): List<RelayContactRequest> = buildList {
        for (index in 0 until length()) add(getJSONObject(index).toContactRequest())
    }

    private fun JSONArray.toReceiptList(): List<RelayReceipt> = buildList {
        for (index in 0 until length()) add(getJSONObject(index).toReceipt())
    }

    private fun JSONArray.toReadCursorList(): List<RelayReadCursor> = buildList {
        for (index in 0 until length()) add(getJSONObject(index).toReadCursor())
    }
}

File: src/relay_protocol.rs
use crate::relay_security::RelayPublicBundleEntry;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RelayEnvelope {
    #[serde(rename = "messageId")]
    pub message_id: String,
    #[serde(rename = "conversationId")]
    pub conversation_id: String,
    #[serde(rename = "senderHandle")]
    pub sender_handle: String,
    #[serde(rename = "recipientHandle")]
    pub recipient_handle: String,
    #[serde(rename = "sessionId")]
    pub session_id: String,
    #[serde(rename = "ciphertextBase64")]
    pub ciphertext_base64: String,
    pub algorithm: String,
    #[serde(rename = "sentAt")]
    pub sent_at: u64,
    #[serde(rename = "senderDeviceId", default)]
    pub sender_device_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RelayContactRequest {
    #[serde(rename = "requestId")]
    pub request_id: String,
    #[serde(rename = "senderHandle")]
    pub sender_handle: String,
    #[serde(rename = "recipientHandle")]
    pub recipient_handle: String,
    #[serde(rename = "senderDisplayName")]
    pub sender_display_name: String,
    #[serde(rename = "publicBundleBase64")]
    pub public_bundle_base64: String,
    #[serde(rename = "identityFingerprint")]
    pub identity_fingerprint: String,
    #[serde(rename = "sentAt")]
    pub sent_at: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RelayReceipt {
    #[serde(rename = "receiptId")]
    pub receipt_id: String,
    #[serde(rename = "messageId")]
    pub message_id: String,
    #[serde(rename = "conversationId")]
    pub conversation_id: String,
    #[serde(rename = "senderHandle")]
    pub sender_handle: String,
    #[serde(rename = "recipientHandle")]
    pub recipient_handle: String,
    #[serde(rename = "recipientDeviceId")]
    pub recipient_device_id: String,
    #[serde(rename = "receiptType")]
    pub receipt_type: String,
    #[serde(rename = "recordedAt")]
    pub recorded_at: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RelayReadCursor {
    #[serde(rename = "cursorId")]
    pub cursor_id: String,
    #[serde(rename = "conversationId")]
    pub conversation_id: String,
    pub handle: String,
    #[serde(rename = "deviceId")]
    pub device_id: String,
    #[serde(rename = "readThroughTimestamp")]
    pub read_through_timestamp: u64,
    #[serde(rename = "recordedAt")]
    pub recorded_at: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum RelayFrame {
    #[serde(rename = "hello")]
    Hello {
        handle: String,
        #[serde(rename = "deviceId")]
        device_id: String,
        #[serde(rename = "displayName")]
        display_name: String,
        #[serde(rename = "publicBundleBase64")]
        public_bundle_base64: String,
        #[serde(rename = "identityFingerprint")]
        identity_fingerprint: String,
    },
    #[serde(rename = "challenge")]
    Challenge {
        #[serde(rename = "relaySessionId")]
        relay_session_id: String,
        challenge: String,
    },
    #[serde(rename = "authenticate")]
    Authenticate {
        handle: String,
        #[serde(rename = "relaySessionId")]
        relay_session_id: String,
        challenge: String,
        #[serde(rename = "publicBundleBase64")]
        public_bundle_base64: String,
        #[serde(rename = "identityFingerprint")]
        identity_fingerprint: String,
        #[serde(rename = "signatureBase64")]
        signature_base64: String,
    },
    #[serde(rename = "authenticated")]
    Authenticated {
        #[serde(rename = "relaySessionId")]
        relay_session_id: String,
        handle: String,
    },
    #[serde(rename = "binding_conflict")]
    BindingConflict {
        handle: String,
        #[serde(rename = "deviceId")]
        device_id: String,
        #[serde(rename = "existingIdentityFingerprint")]
        existing_identity_fingerprint: String,
        #[serde(rename = "requestedIdentityFingerprint")]
        requested_identity_fingerprint: String,
        #[serde(rename = "relinkToken")]
        relink_token: String,
        message: String,
    },
    #[serde(rename = "key_rotation_request")]
    KeyRotationRequest {
        handle: String,
        #[serde(rename = "deviceId")]
        device_id: String,
        #[serde(rename = "currentIdentityFingerprint")]
        current_identity_fingerprint: String,
        #[serde(rename = "newPublicBundleBase64")]
        new_public_bundle_base64: String,
        #[serde(rename = "newIdentityFingerprint")]
        new_identity_fingerprint: String,
    },
    #[serde(rename = "key_rotation_applied")]
    KeyRotationApplied {
        handle: String,
        #[serde(rename = "deviceId")]
        device_id: String,
        #[serde(rename = "oldIdentityFingerprint")]
        old_identity_fingerprint: String,
        #[serde(rename = "newIdentityFingerprint")]
        new_identity_fingerprint: String,
        #[serde(rename = "rotatedAt")]
        rotated_at: u64,
    },
    #[serde(rename = "approve_device_relink")]
    ApproveDeviceRelink {
        handle: String,
        #[serde(rename = "deviceId")]
        device_id: String,
        #[serde(rename = "relinkToken")]
        relink_token: String,
    },
    #[serde(rename = "device_relink_applied")]
    DeviceRelinkApplied {
        handle: String,
        #[serde(rename = "deviceId")]
        device_id: String,
        #[serde(rename = "oldIdentityFingerprint")]
        old_identity_fingerprint: String,
        #[serde(rename = "newIdentityFingerprint")]
        new_identity_fingerprint: String,
        #[serde(rename = "relinkToken")]
        relink_token: String,
        #[serde(rename = "relinkedAt")]
        relinked_at: u64,
    },
    #[serde(rename = "publish")]
    Publish { envelope: RelayEnvelope },
    #[serde(rename = "envelope")]
    Envelope { envelope: RelayEnvelope },
    #[serde(rename = "contact_request")]
    ContactRequest { request: RelayContactRequest },
    #[serde(rename = "receipt")]
    Receipt { receipt: RelayReceipt },
    #[serde(rename = "read_cursor")]
    ReadCursor { cursor: RelayReadCursor },
    #[serde(rename = "delivery_ack")]
    DeliveryAck {
        #[serde(rename = "messageId")]
        message_id: String,
        #[serde(rename = "deliveredAt")]
        delivered_at: u64,
    },
    #[serde(rename = "peer_bundle_request")]
    PeerBundleRequest {
        #[serde(rename = "peerHandle")]
        peer_handle: String,
        #[serde(rename = "deviceId", default, skip_serializing_if = "Option::is_none")]
        device_id: Option<String>,
    },
    #[serde(rename = "peer_bundle_response")]
    PeerBundleResponse {
        #[serde(rename = "peerHandle")]
        peer_handle: String,
        #[serde(rename = "publicBundleBase64")]
        public_bundle_base64: Option<String>,
        #[serde(default)]
        bundles: Vec<RelayPublicBundleEntry>,
    },
    #[serde(rename = "history_sync_request")]
    HistorySyncRequest { since: u64 },
    #[serde(rename = "history_sync_response")]
    HistorySyncResponse {
        #[serde(rename = "relaySessionId")]
        relay_session_id: String,
        #[serde(rename = "syncedUntil")]
        synced_until: u64,
        envelopes: Vec<RelayEnvelope>,
        #[serde(rename = "contactRequests")]
        contact_requests: Vec<RelayContactRequest>,
        receipts: Vec<RelayReceipt>,
        #[serde(rename = "readCursors")]
        read_cursors: Vec<RelayReadCursor>,
    },
    /// WebRTC signaling frame — relayed to the target handle for P2P bootstrap.
    #[serde(rename = "signaling")]
    Signaling {
        #[serde(rename = "fromHandle")]
        from_handle: String,
        #[serde(rename = "toHandle")]
        to_handle: String,
        #[serde(rename = "signalType")]
        signal_type: String,
        payload: String,
        #[serde(rename = "sentAt")]
        sent_at: u64,
    },
    #[serde(rename = "error")]
    Error { message: String },
}
